Implement Series.sum() in new style

PokhodenkoSA opened the pull request:
Implemented Series.sum() via numpy.nansum().
Is it correct solution or should I implement it in another way?

shssf commented on hpat/datatypes/hpat_pandas_series_functions.py:
```diff
@@ -614,6 +615,57 @@ def hpat_pandas_series_sub_number_impl(self, other):
+        """
+        Test: python -m hpat.runtests hpat.tests.test_series.TestSeries.test_series_sum1
+        """
+        return numpy.nansum(self._data)
```
It looks like it should be `numpy.sum` if `skipna` parameter is false

PokhodenkoSA replied:
Done

shssf commented on hpat/datatypes/hpat_pandas_series_functions.py:
```diff
@@ -614,6 +615,57 @@ def hpat_pandas_series_sub_number_impl(self, other):
+        input arg
+    axis:
+        *unsupported*
+    skipna:
```
it looks like you can support `skipna` because you can use Numpy functions for each parameter value

shssf commented on hpat/datatypes/hpat_pandas_series_functions.py:
```diff
@@ -614,6 +615,57 @@ def hpat_pandas_series_sub_number_impl(self, other):
 
 
+@overload_method(SeriesType, 'sum')
+def hpat_pandas_series_sum(self, axis=None, skipna=None, level=None, numeric_only=None, min_count=0, **kwargs):
```
no need to put `kwargs` here
Also, it looks like inside Pandas sources, `skipna` parameter has default `true` (skipna=True). It might be a mistake in Pandas documentation.

PokhodenkoSA replied:
Inside Pandas:
```
if (skipna == None):
    skipna = True
````

shssf commented on hpat/datatypes/hpat_pandas_series_functions.py:
```diff
@@ -614,6 +615,57 @@ def hpat_pandas_series_sub_number_impl(self, other):
+        raise TypingError('{} Unsupported parameters. Given axis: {}, skipna: {}, level: {}, numeric_only: {}, min_count {}'
+                          .format(_func_name, axis, skipna, level, numeric_only, min_count))
+
+    def hpat_pandas_series_sum_impl(self):
```
parameters for this function should be the same as main function

shssf commented on hpat/hiframes/series_kernels.py:
```diff
@@ -492,7 +492,6 @@ def gt_f(a, b):
 
 
 series_replace_funcs = {
-    'sum': _column_sum_impl_basic,
```
please, don't delete it - keep it as comments. It might be helpful for us to see this code in short term. We delete it a bit later

shssf commented on hpat/datatypes/hpat_pandas_series_functions.py:
```diff
@@ -64,6 +64,7 @@
 Implemented methods:
     append
     ne
+    sum
```
no need this here because this doc string is deprecated.

PokhodenkoSA commented:
With HPAT_CONFIG_PIPELINE_HPAT=0 the old test hpat.tests.test_series.TestSeries.test_series_sum2 do not work:
```
No definition for lowering <built-in function add>(series(float64, array(float64, 1d, C), none, False), series(float64, array(float64, 1d, C), none, False)) -> series(float64, array(float64, 1d, C), none, False)

File "hpat\tests\test_series.py", line 859:
        def test_impl(S):
            return (S + S).sum()
            ^

[1] During: lowering "$0.3 = S + S" at c:\Users\spokhode\Documents\hpat-workspace\hpat\hpat\tests\test_series.py (859)
```
As I can understand `add` operator for Series is not implemented yet. @shssf should I fix it in this PR or in new one?

fschlimb commented on hpat/datatypes/hpat_pandas_series_functions.py:
```diff
@@ -614,6 +616,85 @@ def hpat_pandas_series_sub_number_impl(self, other):
+            skipna = True
+        if skipna:
+            return numpy.nansum(self._data)
+        return numpy.sum(self._data)
```
Will this eventually get expanded into a parfor?

PokhodenkoSA replied:
Do not know yet. I will check.

shssf replied:
@fschlimb This call will be inlined from Numba code. We will have no `call` to Numpy here. So, palatalization is questionable here.

shssf commented:
@PokhodenkoSA I see the test (`return (S + S).sum()`) is testing two diffrent things: function `sum()` and `operator+`
This test should be disabled until we implement all required functionality. Example in PR #189 (`@unittest.skipIf(hpat.config.config_pipeline_hpat_default` can be used)
Please create new tests for particular `sum()`

Operators should be implemented separately (in other PRs).

shssf commented:
please resolve conflicts

PokhodenkoSA commented:
@shssf, I need help.

I have made all modifications we usually do for disabling old approach for `sum`:
1. Implement @overload_method in hpat/datatypes/hpat_pandas_series_functions.py
2. Removed `sum` from list in _run_call_series() in hpat/hiframes/hiframes_typed.py
3. Disable adding `resolve_sum` to `SeriesAttribute` in hpat/hiframes/pd_series_ext.py
4. Commented/uncommented `sum` in hpat/hiframes/series_kernels.py:
```
series_replace_funcs = {
    # 'sum': _column_sum_impl_basic,
```
Moreover, I have fixed call for hpat.hiframes.pd_series_ext.SeriesAttribute.resolve_sum() in hpat/hiframes/pd_dataframe_ext.py (SeriesAttribute has no resolve_sum, see point 3 above).

Finally I got this errors:
```
ERROR: test_str_contains_noregex (hpat.tests.test_hiframes.TestHiFrames)
  File "...\AppData\Local\Continuum\anaconda3\envs\hpat-sum-env\lib\site-packages\numba\datamodel\models.py", line 696, in getter
    raise TypeError("expecting {0} but got {1}".format(*args))
TypeError: expecting {i8*, i8*, i64, i64, i8*, [1 x i64], [1 x i64]} but got {{i8*, i8*, i64, i64, i8*, [1 x i64], [1 x i64]}, i8*, i8*}

ERROR: test_str_contains_regex (hpat.tests.test_hiframes.TestHiFrames)
  File "C:\Users\spokhode\AppData\Local\Continuum\anaconda3\envs\hpat-sum-env\lib\site-packages\numba\datamodel\models.py", line 696, in getter
    raise TypeError("expecting {0} but got {1}".format(*args))
TypeError: expecting {i8*, i8*, i64, i64, i8*, [1 x i64], [1 x i64]} but got {{i8*, i8*, i64, i64, i8*, [1 x i64], [1 x i64]}, i8*, i8*}
```
And many errors like that in different test suites:
```
======================================================================
FAIL: test_column_distribution (hpat.tests.test_hiframes.TestHiFrames)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "C:\Users\spokhode\Documents\hpat-workspace\sum-workspace\hpat\hpat\tests\test_hiframes.py", line 161, in test_column_distribution
    self.assertEqual(count_array_REPs(), 0)
AssertionError: 5 != 0
```

I have no idea what to do else.
@shssf, Have you any suggestion what to try or what could be wrong?

shssf commented:
@PokhodenkoSA I will take a look

PokhodenkoSA commented:
I have found solution (thanks to @kozlov-alexey).
I have returned `sum` to list in `_run_call_series()` in `hpat/hiframes/hiframes_typed.py` (step 2)
and uncommented 'sum' in `hpat/hiframes/series_kernels.py` (step 4).
Also I have skipped test for `skipna` because old style does not allow parameters for `sum()`.

shssf commented:
@PokhodenkoSA I started looking for this PR today and didn't understand - everything is working.
later I found you made a changes which make this PR working :-) It is good!
Thank you for squashing commits 👍

PokhodenkoSA commented:
@shssf Thank you for merging it. BTW, GitHub can squash commits before merge https://github.blog/2016-04-01-squash-your-commits/.
